ThunkLibs/libX11/libX11_Host.cpp

```diff
@@ -25,103 +25,301 @@ desc: Handles callbacks and varargs
 #include <utility>
 
 #include "ldr_ptrs.inl"
+#ifdef _M_ARM_64
+__attribute__((naked))
+void* InternalVariadic(void* Ptr, uint64_t a_0, size_t count, void** list) {
+  // Arguments 0-7 is in registers
+  // 8+ stored on to stack
+  // 0 need to be on the end of the list
+  //
+  // Incoming:
+  // x0 = Function
+  // x1 = XIM
+  // x2 = count
+  // x3 = stack array of 64-bit values
+  //
+  // Outgoing:
+  // x0: Ptr
+  __asm volatile(R"(
+    # Store LR and frame pointer if it exists
+    stp x28, x30, [sp, -16]!
+
+    # Move our function to x8, which will be unused
+    mov x8, x0
+
+    # Move XIM to arg0
+    mov x0, x1
+
+    # Move our list to x9, which will be unused
+    mov x9, x3
+
+    # >6 means use stack callback
+    cmp x2, 6
+    b.gt .stack
+
+    # Setup a jump table
+    adr x10, .zero
+    adr x11, .jump_table
+    ldrb w11, [x11, x2]
+    add x10, x10, x11, lsl 2
```

Comment:
```suggestion
    adr x11, .jump_table
    ldrb w11, [x11, x2]
    mov x10, x11, lsl 2
```

Reply:
I'm not sure what you're suggesting here.
We use `adr x10, .zero` to load the base offset address of the jump targets.
Then we index the jump table with `ldrb w11, [x11, x2]`, only needing a byte here since the offsets are so small.
We then add that offset from the jump table to the base of the jump targets, multiplying by two since we compacted with `add <destination reg>, <jump target base>, <jump target offset>, lsl 2`

Reply:
The initialization of x10 (`adr x10, .zero`) and the subsequent `add x10, x10, x11, lsl 2` can be combined into `mov x10, x11, lsl 2`.

In terms of equivalent C code, instead of doing `x10 = 0; x10 = x10 + x11 << 2`, we'd just be doing `x10 = x11 << 2`.

Reply:
`.zero` is a relative address label called `.zero` in the following asm. Not constant zero.

Reply:
so `adr x10, .zero` is effectively going to set x10 to `PC + <N number of instructions>`

Reply:
Derp, you're right of course. Appreciate your patience on this one 🙈